lib/id3c/api/routes.py

```diff
@@ -283,6 +285,121 @@ def get_identifier_set_uses(*, session):
+
+    POST /warehouse/sample to validate, then create new or update
+    existing sample that passes validation.
+
+    JSON object with status (inserted, updated, or failed validation)
+    of the sample is returned.
+    """
+
+    sample = request.get_json()
+
+    LOG.debug(f"Creating/updating sample «{sample}»")
+
+    # everything except _provenance, which is auto-calculated and added below
+    schema = {
+        "type": "object",
+        "properties": {
+            "sample_id": {
+                "type": "string",
+                "minLength": 8,
+                "maxLength": 8
+            },
+            "collection_id": {
+                "type": "string",
+                "minLength": 8,
+                "maxLength": 8
+            },
+            "collection_date": {
+                "type": "string",
+                "format": "date"
+            },
+            "sample_origin": {
+                "type": "string"
+            },
+            "swab_site": {
+                "type": "string"
+            },
+            "clia_id": {
+                "type": "string",
+                "minLength": 8,
+                "maxLength": 8
+            },
+            "received_date": {
+                "type": "string",
+                "format": "date"
+            },
+            "aliquot_a": {
+                "type": "string"
+            },
+            "aliquot_b": {
+                "type": "string"
+            },
+            "aliquoted_date": {
+                "type": "string",
+                "format": "date"
+            },
+            "rack_a": {
+                "type": "string"
+            },
+            "rack_a_nickname": {
+                "type": "string"
+            },
+            "rack_b": {
+                "type": "string"
+            },
+            "rack_b_nickname": {
+                "type": "string"
+            },
+            "swab_type": {
+                "type": "string",
+                "enum": ["ans", "mtb", "np", "tiny", "unk", "none"]
+            },
+            "collection_matrix": {
+                "type": "string",
+                "enum": ["dry", "utm_vtm", "pbs", "none"]
+            },
+            "notes": {
+                "type": "string"
+            }
+        },
+        "anyOf": [
+            { "required":
+                [ "sample" ] },
+            { "required":
+                [ "collection" ] }
+        ]
+    }
+
+    # using DraftXValidator(schema).validate(...) instead of jsonschema.validate(...) to
+    # return accurate error message for "anyOf" requirement
+    try:
+        Draft7Validator(schema).validate(sample)
+    except ValidationError as e:
+        return str(e), 400
+
+    provenance = {
+        "source": request.referrer or request.remote_addr or "?",
+        "method": request.method,
+        "path": request.path,
+        "timestamp": datetime.now(),
+        "sha1sum": sha1(json.dumps(sample).encode()).hexdigest()
+    }
```

Comment:
The `_provenance` for data coming from an API is going to look different from what is currently being used, which is specific to data coming from aliquoting worksheets:
```
"_provenance": {
        "row": 1234,
        "sheet": "aliquoting",
        "sha1sum": "a9993e364706816aba3e25717850c26c9cd0d89d",
        "workbook": "s3://bucket/folder/filename.xlsx"
}
```
Since we'll no longer have a row, sheet, or workbook, I replaces these with source (referrer or IP), method (i.e. "POST"), the endpoint path ("/v1/warehouse/sample") and current timestamp. The sha1sum will be generated per record instead of an entire file.

This seems like a relatively safe thing to change, but there could be some impacts downstream I'm not aware of. One example I found is in `shipping.reportable_condition_v1` view which parses workbook and sheet into columns. That view could be updated to accommodate this new provenance format, but wanted to get some input and finalize the format here before working on related changes.